components/shared/layout/Searchbox.tsx

```diff
@@ -0,0 +1,62 @@
+"use client";
+import { formUrlQuery, removeKeysFromQuery } from "@/lib/utils";
+import { usePathname, useRouter, useSearchParams } from "next/navigation";
+import React, { useEffect, useState } from "react";
+import { Input } from "../../ui/input";
+import Image from "next/image";
+import searchIcon from "@/public/icons/search.svg";
+import cmdKIcon from "@/public/icons/cmd-k.svg";
+interface Props {
+  filters: {
+    name: string;
+    value: string;
+  }[];
+  otherClasses?: string | undefined;
+  containerClasses?: string;
+}
+const Searchbox = ({ filters, otherClasses, containerClasses }: Props) => {
+  // modeled after my searchbox from devoverflow
+  const router = useRouter();
+  const searchParams = useSearchParams();
+  const pathname = usePathname();
+  const query = searchParams.get("q");
+  const [search, setSearch] = useState(query || "");
+  const [isOpen, setIsOpen] = useState(false);
+
+  useEffect(() => {
+    const delayDebounceFn = setTimeout(() => {
+      if (search) {
+        const newUrl = formUrlQuery({
+          params: searchParams.toString(),
+          key: "global",
+          value: search,
+        });
+
+        router.push(newUrl, { scroll: false });
+      } else {
+        if (query) {
+          const newUrl = removeKeysFromQuery({
+            params: searchParams.toString(),
+            keysToRemove: ["global", "type"],
+          });
+
+          router.push(newUrl, { scroll: false });
+        }
+      }
+    }, 300);
+
+    return () => clearTimeout(delayDebounceFn);
+  }, [search, router, pathname, searchParams, query]);
```

Comment:
While I understand this is a work in progress- I would much rather we create a debounce hook (we may want other debounced searching or filtering options for the rest of the project) than to create each debounce as a timer manually when we need it.


I would create a `hooks` folder in our `lib` folder, and create a `useDebounce.tsx` file

```js
// useDebounce.tsx
import { useState, useEffect } from "react";

function useDebounce<T>(value: T, delay: number) {
  const [debouncedValue, setDebouncedValue] = useState<T>(value);

  useEffect(() => {
    const handler = setTimeout(() => {
      setDebouncedValue(value);
    }, delay);

    return () => {
      clearTimeout(handler);
    };
  }, [value, delay]);

  return debouncedValue;
}

export default useDebounce;
```

Here I've made it generic by using `<T>` - which lets us type the values properly and helps TS keep up with what kind of value could be returned from our hook.

Then we can rewrite a little bit of our useEffect:
```js
// SearchBox.tsx
const debouncedSearch = useDebounce<string>(search, 300);

  useEffect(() => {
    if (debouncedSearch) {
      const newUrl = formUrlQuery({
        params: searchParams.toString(),
        key: "q",
        value: debouncedSearch,
      });

      router.push(newUrl, { scroll: false });
    } else {
      if (query) {
        const newUrl = removeKeysFromQuery({
          params: searchParams.toString(),
          keysToRemove: ["q"],
        });

        router.push(newUrl, { scroll: false });
      }
    }
  }, [debouncedSearch, router, pathname, searchParams, query]);
  ```
Now we can easily add debouncing across the rest of the app.

Reply:
noted, ty for th suggestion